#1935: Stunning and Cuffing

JaytleBee opened the pull request:
### Purpose
Implement #1935

### Open Questions and Pre-Merge TODOs

- [x]  This fix is tested on the same branch it is PR'ed to.
- [x]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [x]  This PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

### Notes:
Still needs some work (apparently it can't automatically merge? I'll need to look into that!). This is to be seen as more of a progress report.

JaytleBee committed:
WIP: Adds Stun Baton and fixes some things relating to the tazer.

JaytleBee committed:
Merge remote-tracking branch 'origin/develop' into iss1935-stun-and-cuff

# Conflicts:
#	UnityProject/Assets/Scripts/Input System/Triggers/MeleeTrigger.cs
#	UnityProject/Assets/Scripts/Weapons/Gun.cs

JaytleBee committed:
WIP: Implements issue 1935 (Stunning and Cuffing)

TODO:
- Fix being able to handcuff people even if they move during cuffing (probably add that as a feature to progress bars if we haven't already)
- Pretty things up

JaytleBee committed:
Merge branch 'develop' into iss1935-stun-and-cuff

Also pretties up some things and rewrites cuffs in IF2

TODO: Rewrite stun baton in IF2

JaytleBee committed:
Rewrites Stun Batons in IF2. Removes MIT.

Other changes:
- Adds a lerp effect when stunning someone on help intent
- Changes some things with WeaponNetworkActions to do that
- Adds more comments
- Rewrites a bit of code to be prettier

Jibbajabbafic commented on UnityProject/Assets/Scripts/Input System/Triggers/MeleeTrigger.cs.orig:
```diff
@@ -0,0 +1,100 @@
+﻿using System;
```
Is this file needed anymore?

I'd remove it and the meta file if not

JaytleBee replied:
My bad, that one must have slipped through

JaytleBee committed:
Merge branch 'develop' into iss1935-stun-and-cuff

Also fixes a small bug where clicking cuffs gives an NRE

JaytleBee committed:
Adds handcuffs correctly

Jibbajabbafic commented on UnityProject/Assets/Scripts/Items/Others/Restraint.cs:
```diff
@@ -0,0 +1,59 @@
+
+	protected override bool WillInteract(HandApply interaction, NetworkSide side)
+	{
+		PlayerMove targetPM = interaction.TargetObject?.GetComponent<PlayerMove>();
```
This will handle all the default interact checks (is player in crit state, in range... etc)
```suggestion
		if (!base.WillInteract(interaction, side)) return false;
		
		PlayerMove targetPM = interaction.TargetObject?.GetComponent<PlayerMove>();
```

Jibbajabbafic commented on UnityProject/Assets/Scripts/Player/PlayerMove.cs:
```diff
@@ -317,47 +323,107 @@ public void Unrestrain()
+	}
+
+	/// <summary>
+	/// Called when a player 
```
Called when a player...

Jibbajabbafic commented on UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs:
```diff
@@ -0,0 +1,66 @@
+	/// <summary>
+	/// How long to stun for (in seconds)
+	/// </summary>
+	public float stunTime;
```
If this is public to set it in the editor I'd suggest doing this instead to make it more obvious:
```suggestion
	[SerializeField]
	private float stunTime;
```

Jibbajabbafic commented on UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs:
```diff
@@ -0,0 +1,66 @@
+
+	protected override bool WillInteract(HandApply interaction, NetworkSide side)
+	{
+		return interaction.UsedObject == gameObject
```
```suggestion
		if (!base.WillInteract(interaction, side)) return false;

		return interaction.UsedObject == gameObject
```

Jibbajabbafic commented on UnityProject/Assets/Scripts/Weapons/Melee/StunBaton.cs:
```diff
@@ -0,0 +1,61 @@
+	/// <summary>
+	/// Sound played when turning this baton on/off
+	/// </summary>
+	public string soundToggle;
```
All these don't need to be public
```suggestion
	[SerializeField]
	private string soundToggle;
```

Jibbajabbafic commented on UnityProject/Assets/Scripts/Weapons/Melee/StunBatonActivate.cs:
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections;
```
I think I'd prefer it if this was merged with `StunBaton.cs` so it's all self-contained, but it's fine

JaytleBee replied:
I actually did it like that but StunBaton has to be a network behaviour (because it uses a syncvar) and StunBatonActivate is an Interaction. There's probably a way to get it all in one file but I don't think it's all that important

Jibbajabbafic replied:
You can use `NB(InteractionType)Interactable`, e.g. `NBHandActivateInteractable` 😄

Jibbajabbafic reviewed (commented):
Code looks pretty good.

I have some small suggestions

JaytleBee committed:
Update UnityProject/Assets/Scripts/Items/Others/Restraint.cs

Co-Authored-By: Jibbajabbafic <Jibbajabbafic@users.noreply.github.com>

JaytleBee committed:
Beautifies the code (thanks Jibby!)

- Removes MeleeTrigger.cs.orig & meta file (sorry about that brain fart)
* Merges StunBaton.cs with StunBatonActivate.cs
* Changes public fields that don't need to be public to private fields
with [SerializeField] (and public getters where necessary)

JaytleBee committed:
Update UnityProject/Assets/Scripts/Weapons/Melee/MeleeStun.cs

Co-Authored-By: Jibbajabbafic <Jibbajabbafic@users.noreply.github.com>

JaytleBee committed:
Fixes sounds for cuffing and stunning

JaytleBee committed:
Fixes Handcuff sound and prefabs

* Sets position of prefabs to 0
* Sets scale of stun baton to 0
* Handcuffs now correctly play their sound when starting to cuff (again)
